File: app/src/main/java/com/robotnec/budget/fragments/AddAccountFragment.java
package com.robotnec.budget.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;

import butterknife.Bind;
import butterknife.OnClick;
import com.robotnec.budget.R;
import com.robotnec.budget.core.domain.Account;
import com.robotnec.budget.core.mvp.presenter.AddAccountPresenter;
import com.robotnec.budget.core.mvp.view.AddAccountView;
import com.robotnec.budget.util.Keys;

/**
 * @author zak <zak@swingpulse.com>
 */
public class AddAccountFragment extends BasePresenterFragment<AddAccountPresenter>
        implements AddAccountView {

    @Bind(R.id.toolbar)
    Toolbar mToolbar;

    @Bind(R.id.edit_account_name)
    EditText mEditAccountName;

    @Bind(R.id.edit_account_initial_amount)
    EditText mEditAccountInitialAmount;

    @Bind(R.id.spinner_currency)
    Spinner mSpinnerAccountCurrency;

    @Bind(R.id.button_delete)
    Button mButtonDelete;

    public static AddAccountFragment newInstance(Account account) {
        AddAccountFragment fragment = new AddAccountFragment();
        Bundle args = new Bundle();
        args.putSerializable(Keys.ACCOUNT, account);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_add_account;
    }

    @Override
    protected AddAccountPresenter createPresenter() {
        Account account = (Account) getArguments().getSerializable(Keys.ACCOUNT);
        return new AddAccountPresenter(this, account);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initToolbarBack(mToolbar);
    }

    @OnClick(R.id.button_delete)
    void onDeleteClick() {
        mPresenter.deleteAccount();
    }

    @OnClick(R.id.button_done)
    void onDoneClick() {
        String accountName = mEditAccountName.getText().toString();
        String accountInitialAmount = mEditAccountInitialAmount.getText().toString();
        String accountCurrency = mSpinnerAccountCurrency.getSelectedItem().toString();
        mPresenter.addOrEditAccount(accountName, accountInitialAmount, accountCurrency);
        getActivity().finish();
    }

    @Override
    public void initEditMode(Account account) {
        mEditAccountName.setText(account.getName());
        mEditAccountName.setSelection(mEditAccountName.length());

        mEditAccountInitialAmount.setText(String.valueOf(account.getAmount()));
        mEditAccountInitialAmount.setSelection(mEditAccountInitialAmount.length());

        mButtonDelete.setVisibility(View.VISIBLE);

        // TODO set currency spinner position
    }

    @Override
    public void closeView() {
        getActivity().finish();
    }
}

File: app/src/main/java/com/robotnec/budget/app/fragments/AddAccountFragment.kt
package com.robotnec.budget.app.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View
import android.widget.ArrayAdapter
import com.robotnec.budget.R
import com.robotnec.budget.core.domain.Account
import com.robotnec.budget.core.mvp.presenter.AddAccountPresenter
import com.robotnec.budget.core.mvp.view.AddAccountView
import kotlinx.android.synthetic.main.fragment_add_account.buttonDelete
import kotlinx.android.synthetic.main.fragment_add_account.buttonDone
import kotlinx.android.synthetic.main.fragment_add_account.editAccountInitialAmount
import kotlinx.android.synthetic.main.fragment_add_account.editAccountName
import kotlinx.android.synthetic.main.fragment_add_account.spinnerCurrency
import kotlinx.android.synthetic.main.fragment_add_account.toolbar
import org.jetbrains.anko.support.v4.withArguments
import org.joda.money.CurrencyUnit

/**
 * @author zak zak@swingpulse.com>
 */
class AddAccountFragment : BasePresenterFragment<AddAccountPresenter>(), AddAccountView {

    companion object {
        const val ARG_ACCOUNT = "account"

        fun newInstance(account: Account?): Fragment {
            if (account != null) {
                return AddAccountFragment().withArguments(ARG_ACCOUNT to account)
            } else {
                return AddAccountFragment().withArguments()
            }
        }
    }

    override val layoutId: Int
        get() = R.layout.fragment_add_account

    override fun createPresenter(): AddAccountPresenter {
        val account = arguments[ARG_ACCOUNT] as? Account
        return AddAccountPresenter(this, account)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initToolbarBack(toolbar)
        val adapter = ArrayAdapter(context,
                android.R.layout.simple_spinner_dropdown_item,
                CurrencyUnit.registeredCurrencies())
        spinnerCurrency.adapter = adapter

        buttonDelete.setOnClickListener { presenter.deleteAccount() }
        buttonDone.setOnClickListener {
            val accountName = editAccountName.text.toString()
            val accountInitialAmount = editAccountInitialAmount.text.toString()
            val accountCurrency = spinnerCurrency.selectedItem.toString()
            presenter.addOrEditAccount(accountName, accountInitialAmount, accountCurrency)
            activity.finish()
        }
    }

    override fun initEditMode(account: Account) {
        editAccountName.setText(account.name)
        editAccountName.setSelection(editAccountName.length())

        editAccountInitialAmount.setText(account.amount.toString())
        editAccountInitialAmount.setSelection(editAccountInitialAmount.length())

        buttonDelete.visibility = View.VISIBLE

        // TODO set currency spinner position
    }

    override fun closeView() {
        activity.finish()
    }
}
